### DISCUSSION TITLE
http middleware for converting connect http request to grpc request

### DISCUSSION THREAD

**Discussion by rauanmayemir** [connectrpc/connect-go#787] [Q&A] [answered]
I'm building a gRPC-server as a mux handler that is supposed to chain via http middlewares and then ultimately apply `grpcServer.ServeHTTP(writer, req)`.

What is a proper way to set up http middleware for connect converter that does not know anything about grpc/protobuf types (an isolated http handler func)?

Basically, I want to transiently convert inbound http request into a grpc-compliant http request with all appropriate headers, etc.
Docs only provide examples where I'm building a connectrpc server which is not my case.

Seeing the envoy http filter I am sure this is possible as `envoy.filters.http.connect_grpc_bridge` has no knowledge of protobuf types.

**Comment by rauanmayemir** [comment]
Jus discovered https://github.com/connectrpc/vanguard-go, but IIUIC it needs to wrap the grpcServer and what I need should only work with `http.Request`

**Comment by rauanmayemir** [comment]
Vanguard generally looks like a perfect tool for the job. 😍

**Comment by timostamm** [comment, marked as answer]
Hey Rauan, it is correct that it's possible to convert a Connect request to a gRPC request without knowledge of the Protobuf schema - as long as both use the same codec (JSON or Protobuf binary).

This is out of scope for `connect-go`, but Vanguard supports this translation (and a bit more) and should do what you want - `vanguard.Transcoder` is a `http.Handler`. If you have any questions or feature request, it's best to open an issue in the Vanguard repository.

**Reply by rauanmayemir** [reply]
Thanks you, will start experimenting with vanguard! 🙏
